src/workerd/api/r2-bucket.c++

```diff
@@ -797,17 +823,94 @@ jsg::Promise<R2Bucket::ListResult> R2Bucket::list(
+  // => empty string also results in termination
+  // There are 2 recursive calls in this function body, each of them always moves the start of the
+  // condHeader to some value found in the condHeader + 1.
+  // => upon each recursion, the size of condHeader is reduced by at least 1.
+  // Eventually we must arrive at an empty string, hence triggering the stop condition.
+
+  size_t nextWildcard = condHeader.findFirst('*').orDefault(SIZE_MAX);
+  size_t nextQuotation = condHeader.findFirst('"').orDefault(SIZE_MAX);
+  size_t nextWeak = condHeader.findFirst('W').orDefault(SIZE_MAX);
+  size_t nextComma = condHeader.findFirst(',').orDefault(SIZE_MAX);
+
+  if (nextQuotation == SIZE_MAX && nextWildcard == SIZE_MAX) {
+  // Both of these being SIZE_MAX means no more wildcards or double quotes are left in the header.
+  // When this is the case, there's no more useful etags that can potentially still be extracted.
+    return etagAccumulator.releaseAsArray();
+  }
+
+  if (nextComma < nextWildcard && nextComma < nextQuotation && nextComma < nextWeak) {
+  // Get rid of leading commas, this can happen during recursion because servers must deal with
+  // empty list elements. E.g.: If-None-Match "abc", , "cdef" should be accepted by the server.
+    // This slice is always safe, since we're at most setting start to the last index + 1,
+    // which just results in an empty list if it's out of bounds by 1.
+    return parseConditionalEtagHeader(condHeader.slice(nextComma + 1), kj::mv(etagAccumulator));
+  } else if (leadingCommaRequired) {
+  // Did not find a leading comma, and we expected a leading comma before any further etags
+    KJ_FAIL_REQUIRE("COMMA EXPECTED TO SEPARATE ETAGS");
+  }
+
+  if (nextWildcard < nextQuotation) {
+    // Unquoted wildcard found
+    // remove all other etags since they're overridden by the wildcard anyways
+    etagAccumulator.clear();
+    struct R2Bucket::WildcardEtag etag = {};
+    etagAccumulator.add(kj::mv(etag));
+    return etagAccumulator.releaseAsArray();
+  }
+  if (nextQuotation < nextWildcard) {
+    size_t etagValueStart = nextQuotation + 1;
+    // Find closing quotation mark, instead of going by the next comma.
+    // This is done because commas are allowed in etags, and double quotes are not.
+    kj::Maybe<size_t> closingQuotation =
+      condHeader.slice(etagValueStart)
+      .findFirst('"')
+      .map([=](size_t cq) { return cq + etagValueStart; });
+
+    KJ_IF_MAYBE(cq, closingQuotation) {
+      // Slice end is non inclusive, meaning that this drops the closingQuotation from the etag
+      kj::String etagValue = kj::str(condHeader.slice(etagValueStart, *cq));
+      if (nextWeak < nextQuotation) {
+        KJ_REQUIRE(
+          condHeader.size() > nextWeak + 2
+            && condHeader[nextWeak + 1] == '/'
+            && nextWeak + 2 == nextQuotation,
+          "WEAK ETAGS MUST START WITH W/ and their value must be quoted"
```

Comment:
Nit... why the allcaps on the first half of this?

Reply:
Changed to lower case now :)